egs/musdb18/X-UMX/train.py

```diff
@@ -0,0 +1,481 @@
+            time_hat = torch.cat([time_hat, tmpi], dim=0)
+
+        mix_t = mix_ref[: n_batch * n_channel, Ellipsis].repeat(n_src + len(indices), 1)
+        refrences_t = mix_ref[n_batch * n_channel :, Ellipsis]
+    else:
+        mix_t = mix_ref[: n_batch * n_channel, Ellipsis].repeat(n_src, 1)
+        refrences_t = mix_ref[n_batch * n_channel :, Ellipsis]
+
+    # Calculation
+    _loss_sdr = weighted_sdr(time_hat, refrences_t, mix_t)
+
+    return 1.0 + _loss_sdr
+
+
+def weighted_sdr(input, gt, mix, weighted=True, eps=1e-10):
+    # ``input'', ``gt'' and ``mix'' should be (Batch, Time Length)
+    assert input.shape == gt.shape
+    assert mix.shape == gt.shape
+
+    ns = mix - gt
+    ns_hat = mix - input
+
+    if weighted:
+        alpha_num = (gt * gt).sum(1, keepdims=True)
+        alpha_denom = (gt * gt).sum(1, keepdims=True) + (ns * ns).sum(1, keepdims=True)
+        alpha = alpha_num / (alpha_denom + eps)
+    else:
+        alpha = 0.5
+
+    # Target
+    num_cln = (input * gt).sum(1, keepdims=True)
+    denom_cln = torch.sqrt(eps + (input * input).sum(1, keepdims=True)) * torch.sqrt(
+        eps + (gt * gt).sum(1, keepdims=True)
+    )
+    sdr_cln = num_cln / (denom_cln + eps)
+
+    # Noise
+    num_noise = (ns * ns_hat).sum(1, keepdims=True)
+    denom_noise = torch.sqrt(eps + (ns_hat * ns_hat).sum(1, keepdims=True)) * torch.sqrt(
+        eps + (ns * ns).sum(1, keepdims=True)
+    )
+    sdr_noise = num_noise / (denom_noise + eps)
+
+    return torch.mean(-alpha * sdr_cln - (1.0 - alpha) * sdr_noise)
+
+
+class MultiDomainLoss(_Loss):
+    """A class for calculating loss functions of X-UMX.
+
+    Args:
+        window_length (int): The length in samples of window function to use in STFT.
+        in_chan (int): Number of input channels, should be equal to
+            STFT size and STFT window length in samples.
+        n_hop (int): STFT hop length in samples.
+        spec_power (int): Exponent for spectrogram calculation.
+        nb_channels (int): set number of channels for model (1 for mono
+            (spectral downmix is applied,) 2 for stereo).
+        loss_combine_sources (bool): Set to true if you are using the combination scheme
+            proposed in [1]. If you select ``loss_combine_sources: yes'' via
+            conf.yml, this is set as True.
+        loss_use_multidomain (bool): Set to true if you are using a frequency- and time-domain
+            losses collaboratively, i.e., Multi Domain Loss (MDL) proposed in [1].
+            If you select ``loss_use_multidomain: yes'' via conf.yml, this is set as True.
+        mix_coef (float): A mixing parameter for multi domain losses
+
+    References
+        [1] "All for One and One for All: Improving Music Separation by Bridging
+        Networks", Ryosuke Sawata, Stefan Uhlich, Shusuke Takahashi and Yuki Mitsufuji.
+        https://arxiv.org/abs/2010.04228 (and ICASSP 2021)
+    """
+
+    def __init__(self, args):
```

Comment:
Receiving `args` in a class is bad practice. 
You have to read the code to know what it uses, it's not transparent. 

Let's not change it to release faster, but next time, let's avoid this.

Reply:
OK, I understand.
I'm sorry for my inappropriate implementation, and appreciate your advise for the next time!